### ISSUE TITLE
Bulk operations - Ability to perform actions on multiple nodes

### ISSUE THREAD

**Issue by yogendra** [canonical/multipass#2404]
**What are you trying to do?**
Currently all commands use instance name to perform operations. This makes operations on multiple instances difficult and requires individual commands/action on each instance. For example: if I have 3 VMs for apps, 2 vms for db, and I want to run a command on just db nodes, I would have to run two commands. Or If I have to reboot app servers, thats 3 commands. 

**What's your proposed solution?**
One option is to allow adding tags/lables to hosts. While launching we can add add these tags in commands line. 

```
multipass launch --name app-server1 -m2g -c1 -d10g --tags group=app-server
multipass launch --name app-server2 -m2g -c1 -d10g --tags group=app-server
multipass launch --name app-server3 -m2g -c1 -d10g --tags group=app-server

```

Then for any operations we can specify the tag, instead of instance name.

```
multipass stop --tags group=app-server
multipass exec --tags group-app-server exec systemctl daemon-reload
multipass start --tags grou=app-server
```

Additionally, if we can expose these tags to the VM environment (file, metadata, etc.), we can make use of them in config management tool like Chef/Puppet/Script/Cloud-init, etc.

**Additional context**
I made a small utility for multi-node microk8s cluster [yogendra/mnmk8s](https://github.com/yogendra/mnmk8s) (terrible name, i know), that does something similar, in absence of this feature.

**Comment by Saviq** [comment]
Hi @yogendra that's a nice idea indeed :)

Note you can use `--all` on `start`, `stop`, `restart` - that doesn't give you the flexibility of tags of course. As I think of it a tag would likely be just a value, not a key/value pair, but that should achieve the result you're after.

**Comment by yogendra** [comment]
Yes I agree. Label is more appropriate term. Tag tend to attract more complex comparison operations.

Yes I saw --all and it works for actions on all vms. But subset of vm is the issue that I am trying to solve.

**Comment by ricab** [comment]
Thanks, great idea indeed! I hope you don't mind me adding my own two cents :slightly_smiling_face: 

A simpler approach to achieve this would be to use the name to tag instances as needed and then accept either regular expressions or wildcards in commands. For instance, in your case you would be able to do `multipass stop 'app-server.*'` (regex) or `multipass stop 'app-server*'` (wildcard). I think the name allows sufficient freedom that no tag property is required, especially once we allow instance renaming. 

A tangently related proposal that has been on the back of my mind is the ability to prefix instances with the daemon scope (only `local.` for now). This would make existing commands consistent with `get` and `set` in the instance-modification feature (upcoming). IOW, the user *could* do:

```
multipass stop local.foo
multipass set local.foo.cpus=3
multipass start local.foo
```

The `local.` prefix would be optional, and `multipass stop foo` would work just as well (local daemon would be assumed if/when we introduce remote daemons). This is something that started cooking in my mind when I hit "up" in the shell to edit the last command but had to remove/add "local." to identify the same instance, which feels awkward.

Mixing all changes (remote daemons with prefixes + bulk actions) one would be able to do something like `multipass exec *.app-server* -- mycmd` and have `mycmd` executed in all instances with that name, in all daemons, at once. For local app-servers, just `multipass exec app-server* -- mycmd` would work.

For the record, discussing with @Saviq, he proposed wildcards in place of regular expressions. I tend to agree that they would be better for this than regexes (which were what I initially considered). Wildcards are powerful enough for the purpose, but have simpler usage. And they would be consistent with a generalization of our commands to take multiple instance arguments (i.e. `multipass start 'app-server*'` == `multipass start app-server1 app-server2 app-server3`). Quoting may be needed in any case, to prevent shell replacing it.

@Saviq mentioned also `multipass switch` as the way to select the remote to target. I think that would work well with the above if it specified the *default* target. For instance, after `multipass switch bar`, `multipass start foo` would start the instance `foo` in the daemon `bar`, while `multipass start local.foo` would do so in the local daemon.

One final comment is that, when dealing with multiple instances at once, we need to design error handling carefully. Do we want a transactional approach? Is best effort enough? What about timeouts? Do we do this sequentially or in parallel? 

Food for thought. Thanks again for proposing!

**Comment by yogendra** [comment]
@ricab Thanks for a great insight. This should work for use cases that I am thinking. Where can I learn about remote multipass?

**Comment by Saviq** [comment]
@yogendra I just wrote something up on our discourse about it:

https://discourse.ubuntu.com/t/using-multipass-remotely-a-preview/26360

**Comment by ricab** [comment]
Hey @yogendra, I was thinking about a potential future where a multipass client would be able to contact multiple daemons. We have chatted about such possibilities, but have nothing concrete for the time being. Fortunately @Saviq just came up with an alternative that would work today :)

**Comment by adtech97** [comment]
Hello @yogendra. I landed here from a good first issue. The issue seems to have been open for quite a while, and I would like to contribute to this if it is not already closed.

**Comment by vicdeveloperr** [comment]
I'm working on this.

**Comment by vicdeveloperr** [comment]
> Hello [@yogendra](https://github.com/yogendra). I landed here from a good first issue. The issue seems to have been open for quite a while, and I would like to contribute to this if it is not already closed.

Hi, man.
You are already working on this?

**Comment by yogendra** [comment]
@vicdeveloperr - no have not started looking at this. I have ventured into a different space so not looking at this at all.
